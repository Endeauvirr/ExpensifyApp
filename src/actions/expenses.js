import uuidv1 from 'uuid';
import database from '../firebase/firebase';

// export const addExpense = ({
//   description = '',
//   note = '',
//   amount = 0,
//   createdAt = 0 
// } = {}) => ({
//   type: 'ADD_EXPENSE',
//   expense: {
//     id: uuidv1(),
//     description,
//     note,
//     amount,
//     createdAt
//   }
// });

const convertDataToArray = (snapshot) => {
  const expensesArray = [];

  snapshot.forEach((childSnapshot) => {
    expensesArray.push({
      id: childSnapshot.key,
      ...childSnapshot.val()
    });
  });

  return expensesArray;
};

export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0 
    } = expenseData;

    const expense = {
      description,
      note,
      amount,
      createdAt
    };

    return database.ref('expenses').push(expense)
      .then((ref) => {
        dispatch(addExpense({
          id: ref.key,
          ...expense
        }));
      }).catch((error) => {
        console.warn(`startAddExpense action error: ${error}`);
      });
  };
};

export const removeExpense = (id = '') => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = (id = '') => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`)
      .remove()
      .then(() => {
        dispatch(removeExpense(id));
      });
  };
};

export const editExpense = (id, { ...expense } = {}) => ({
  type: 'EDIT_EXPENSE',
  id,
  expense: {
    ...expense
  }
});

export const startEditExpense = (id, { ...expense } = {}) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).update({
      ...expense
    }).then(() => {
      dispatch(editExpense(id, { ...expense }));
    });
  };
};


// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});


// START_SET_EXPENSES (async)

export const startSetExpenses = () => {
  return (dispatch) => {
    return database.ref('expenses')
      .once('value')
      .then((snapshot) => {
        const expenses = convertDataToArray(snapshot);

        dispatch(setExpenses(expenses));
      });
  };
};
