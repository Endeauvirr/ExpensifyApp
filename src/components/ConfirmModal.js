import React from 'react';
import Modal from 'react-modal';

const ConfirmModal = (props) => {
  Modal.setAppElement('body');

  return (
    <Modal
      isOpen={props.isOpen}
      onRequestClose={props.handleCloseModal}
      key={props.id}
      className="modal-confirm"
    >
      <section className="modal__wrapper">
        <header className="modal-header__wrapper">
          <h3 className="modal-header__title">
            Confirm removing expense
          </h3>
        </header>
        <div className="modal-body__wrapper">
          <p>You want to remove <span>{props.description}</span></p>
          <p>Are you sure you want to remove this expense?</p>
        </div>
        <footer className="modal-footer__wrapper">
          <button
            className="modal-footer__cancel-button"
            onClick={props.handleCloseModal}
          >
            Cancel
          </button>
          <button
            className="modal-footer__confirm-button"
            onClick={() => {
              props.handleConfirmModal(props.id);
            }}
          >
            Confirm
          </button>
        </footer>
      </section>
    </Modal>
  );
};

export default ConfirmModal;
