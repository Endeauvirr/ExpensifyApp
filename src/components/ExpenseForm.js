import React from 'react';
import moment from 'moment';

import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';

class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? props.expense.amount : 0,
      note: props.expense ? props.expense.note : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      formValidationError: false
    }
  }
  componentDidMount() {
    if (typeof this.props.expense !== 'undefined') {
      const { description, note, amount, createdAt } = this.props.expense;

      this.setState(() => ({
        description,
        note,
        amount: (amount / 100).toString(),
        createdAt: moment(createdAt)
      }));
    }
  }
  onStringElementChange = (event) => {
    const formFieldName = event.target.name;
    const formFieldValue = event.target.value;

    this.setState(() => ({
      [formFieldName]: formFieldValue
    }))
  }
  onAmountElementChange = (event) => {
    const formFieldName = event.target.name;
    const formFieldValue = event.target.value.replace(',','.');

    if (!formFieldValue || formFieldValue.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({
        [formFieldName]: formFieldValue
      }))
    }
  }
  onAmountElementFocus = (event) => {
    const value = Number(event.target.value);

    if (value === 0) {
      this.setState(() => ({
        amount: ''
      }))
    }
  }
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({
        createdAt
      }));
    }
  }
  onCalendarFocused = ({ focused }) => {
    this.setState(() => ({
      calendarFocused: focused
    }))
  }
  onSubmitHandler = (event) => {
    event.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        formValidationError: true
      }));
    } else {
      this.setState(() => ({
        formValidationError: false
      }));

      const {
        description, amount, createdAt, note
      } = this.state;

      this.props.onSubmit({
        description,
        amount: parseFloat(amount, 10) * 100,
        createdAt: createdAt.valueOf(),
        note
      });
    }
  }
  render() {
    return (
      <div className="expenses__create-expense--form">
        <form onSubmit={this.onSubmitHandler}>
          {this.state.formValidationError && <p>Please fill description and amount fields</p>}
          <input
            type="text"
            name="description"
            placeholder="Description"
            value={this.state.description}
            onInput={this.onStringElementChange}
          />
          <input
            type="text"
            name="amount"
            placeholder="Amount"
            value={this.state.amount}
            onFocus={this.onAmountElementFocus}
            onChange={this.onAmountElementChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onCalendarFocused}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            name="note"
            placeholder="Place a note for expense (optional)"
            onInput={this.onStringElementChange}
          />
          <button
            type="submit"
          >
            {(this.props.expense) ? 'Edit expense' : 'Create expense'}
          </button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
