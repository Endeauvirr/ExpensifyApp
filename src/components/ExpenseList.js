import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
  <div className="expenses__list">
    <h1>Hello im a expense list</h1>
    {
      props.expenses.length === 0 ? (
        <p>No expenses added. Add some more!</p>
      ) : (
        props.expenses.map((expense) => {
          return (
            <ExpenseListItem key={expense.id} {...expense} />
          );
        })
      )
    }
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};


export default connect(mapStateToProps)(ExpenseList);
