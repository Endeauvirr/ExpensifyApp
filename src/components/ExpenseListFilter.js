import React from 'react';
import { connect } from 'react-redux';

import 'react-dates/initialize';
import { DateRangePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilter extends React.Component {
  state = {
    calendarFocused: null
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({
      calendarFocused
    }));
  };
  onTextFilterChange = (event) => {
    this.props.setTextFilter(event.target.value);
  };
  onFilterTypeChange = (event) => {
    if (event.target.value === 'amount') {
      this.props.sortByAmount();
    } else {
      this.props.sortByDate();
    }
  };
  render() {
    return (
      <div className="expenses__list-filters">
        <input
          type="text"
          value={this.props.filters.text}
          onChange={this.onTextFilterChange}
        />

        <select
          name="expensesTypeFilter"
          id="expensesTypeFilter"
          onChange={this.onFilterTypeChange}
        >
          <option value="date" default>by date</option>
          <option value="amount">by amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    );
  };
}

const mapStateToProps = (state) => ({
  filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);
