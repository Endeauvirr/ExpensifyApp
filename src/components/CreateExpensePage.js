import React from 'react';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class CreateExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <h1>Create a new expense</h1>
        <ExpenseForm
          onSubmit={this.onSubmit}
        />
      </div>
    )
  }
};

// Przerabiamy ten bezstanowy komponent na jego klasowy odpowiednik by wyciagnąć metodę onSubmit jako props.
// const CreateExpensePage = (props) => (
//   <div>
//     <h1>Create a new expense</h1>
//     <ExpenseForm
//       onSubmit={(expense) => {
//         // props.dispatch(addExpense(expense));  // Zamieniamy to wywołanie, przy pomocy metody z reduxa mapDispathchToProps() , by łatwiej testować funkcję addExpense()
//         props.onSubmit(expense);
//         props.history.push('/');
//       }}
//     />
//   </div>
// );

const mapDispatchToProps = (dispatch) => {
  return {
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
  };
};

export default connect(undefined, mapDispatchToProps)(CreateExpensePage);
// metoda connect przyjmuje kilka argumentow, undefined na poczatku to argument dla mapStateToProps
