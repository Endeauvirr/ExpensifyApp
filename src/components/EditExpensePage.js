import React from 'react';
import { connect } from 'react-redux';
import { startEditExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  }
  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
      </div>
    );
  }
};

// const EditExpensePage = (props) => {
//   console.log(props);

//   return (
//     <div>
//       This is from edit expense page. Im editing the page with ID of {props.match.params.id}

//       <ExpenseForm
//         expense={props.expense}
//         onSubmit={(expense) => {
//           props.dispatch(editExpense(props.match.params.id, expense));
//           props.history.push('/');
//         }}
//       />
//     </div>
//   );
// };

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense))
  }
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
