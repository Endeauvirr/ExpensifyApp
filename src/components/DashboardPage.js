import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilter';

const DashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default DashboardPage;
