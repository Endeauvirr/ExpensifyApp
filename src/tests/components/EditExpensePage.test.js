import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixture/expenses';

let editExpense;
let history;
let wrapper;

beforeEach(() => {
  editExpense = jest.fn();
  history = { push: jest.fn() };

  wrapper = shallow(<EditExpensePage editExpense={editExpense} history={history} expense={expenses[1]} />);
});

test('Should render EditExpensePage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});


test('Should handle editExpense', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);

  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id, expenses[1]);
});
