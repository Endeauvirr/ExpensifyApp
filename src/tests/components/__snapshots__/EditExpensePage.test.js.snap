// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render EditExpensePage correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 750,
        "createdAt": -345600000,
        "description": "Ball",
        "id": "2",
        "note": "Just simple note 2",
      }
    }
    onSubmit={[Function]}
  />
</div>
`;
